### PR TITLE
Add B.Protocol

### PR DESCRIPTION
**Pull request by yaronvel**
Protocol: https://bprotocol.org

This connector adds support for interacting with Maker and Compound via B.Protocol smart contract. **EDIT:**
the connector also add support to B.Protocol BAMM over liquity protocol stability pool.

Users can do everything the can do with Maker and Compound and get exactly the same conditions, and in addition get to share liquidation proceeds, and when applicable B.Protocol governance tokens.
**EDIT:** User can deposit to liquity's stability pool via B.Protocol and get their account automatically rebalanced whenever liquidation occurs. And this way can get passive yield on their LUSD deposits.

The implementation is a simple clone of the Compound and Maker connectors, with small adaptations to the B.Protocol system.
1. Using different addresses for CDP manager, and comptroller.
2. Adjusting the user debt report, to include the debt that might have been partially paid by the B.Protocol liquidators.

**EDIT:**
`ConnectV2BMakerDAO` Ethereum mainnet address = `0xB0A1f10FeEfECf25064CE7cdF0a65042F7dE7bF0`
`ConnectV2BCompound` Ethereum mainnet address = `0xa3EeFDc2de9DFA59968bEcff3E15b53E6162460f`
`ConnectV2BLiquity` Ethereum mainnet address = `0x19574E5Dfb40bbD63A4F3bdcF27ed662b329b2ff`

### CODE REVIEW

**Review by KaymasJain — APPROVED** [review]

**Comment by thrilok209** [comment]
### Deployed connector:

**Connector Name**: "B-COMPOUND-A"
**Connector Address**: [0xa3EeFDc2de9DFA59968bEcff3E15b53E6162460f](https://etherscan.io/address/0xa3EeFDc2de9DFA59968bEcff3E15b53E6162460f)

**Connector Name**: "B-MAKERDAO-A"
**Connector Address**: [0xB0A1f10FeEfECf25064CE7cdF0a65042F7dE7bF0](https://etherscan.io/address/0xB0A1f10FeEfECf25064CE7cdF0a65042F7dE7bF0)

**Connector Name**: "B-LIQUITY-A"
**Connector Address**: [0x19574E5Dfb40bbD63A4F3bdcF27ed662b329b2ff](https://etherscan.io/address/0x19574E5Dfb40bbD63A4F3bdcF27ed662b329b2ff)

File: contracts/mainnet/connectors/b.protocol/compound/helpers.sol [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+abstract contract Helpers is DSMath, Basic {
+    /**
+     * @dev Compound Comptroller
+     */
+    ComptrollerInterface internal constant troller = ComptrollerInterface(0x9dB10B9429989cC13408d7368644D4A1CB704ea3);
+
+    /**
+     * @dev Compound Mapping
+     */
+    CompoundMappingInterface internal constant compMapping = CompoundMappingInterface(0xA8F9D4aA7319C54C04404765117ddBf9448E2082);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/b.protocol/compound/helpers.sol]:
```suggestion
    CompoundMappingInterface internal constant compMapping = CompoundMappingInterface(0xe7a85d0adDB972A4f0A4e57B698B37f171519e88);
```

File: contracts/mainnet/connectors/b.protocol/compound/main.sol [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        uint cTokenAmt,
+        uint getId,
+        uint setId
+    ) external payable returns (string memory _eventName, bytes memory _eventParam) {
+        (address token, address cToken) = getMapping(tokenId);
+        (_eventName, _eventParam) = withdrawCTokenRaw(token, cToken, cTokenAmt, getId, setId);
+    }
+}
+
+contract ConnectV1BCompound is BCompoundResolver {
```
Comment (thrilok209) [inline contracts/mainnet/connectors/b.protocol/compound/main.sol]:
```suggestion
contract ConnectV2BCompound is BCompoundResolver {
```

File: contracts/mainnet/connectors/b.protocol/makerdao/main.sol [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+        );
+
+        setUint(setId, _amt);
+
+        _eventName = "LogWithdrawDai(uint256,uint256,uint256)";
+        _eventParam = abi.encode(_amt, getId, setId);
+    }
+}
+
+contract ConnectV1BMakerDAO is BMakerResolver {
```
Comment (thrilok209) [inline contracts/mainnet/connectors/b.protocol/makerdao/main.sol]:
```suggestion
contract ConnectV2BMakerDAO is BMakerResolver {
```

**Comment by thrilok209** [comment]
"B-COMPOUND-A",  "B-MAKERDAO-A", "B-LIQUITY-A" connectors whitelisted: https://etherscan.io/tx/0x2ee045e98e449f2d455a303820138617d8d18f86e044be170f9e2c1345233c7f

**Review by thrilok209 — APPROVED** [review]

**Comment by yaronvel** [comment]
@thrilok209 the B-COMPOUND connector you whitelisted does not contains the modified compound mapping address. is it ok?

**Comment by thrilok209** [comment]
@yaronvel new Compound mapping contract is this: https://etherscan.io/address/0xe7a85d0adDB972A4f0A4e57B698B37f171519e88#code

it's updated on B-COMPOUND connector, right? Or Am I missing something?
image

**Comment by yaronvel** [comment]
@thrilok209 a yes, my mistake.
I was not aware you could also edit my comments, and i saw you deployed the same address I wrote at my comments, so I assumed it was my deployment. But now I see it is not the case, and I guess you edited there.

**Pull request merged**